### PR TITLE
Added stream support on hough circles, lines and segments

### PR DESCRIPTION
**Pull request by Atlas42**
- Passed the stream to the different cuda, OpenCV and thurst library calls
- Replace all device by cuda synchronizes
- Added extra synchronize calls after device to host transfers
- Replaced the cuda globals by allocated values

<cut/>

### Pull Request Readiness Checklist

See details at https://github.com/opencv/opencv/wiki/How_to_contribute#making-a-good-pull-request

- [x] I agree to contribute to the project under Apache 2 License.
- [x] To the best of my knowledge, the proposed patch is not based on a code under GPL or other license that is incompatible with OpenCV
- [x] The PR is proposed to proper branch
- [ ] There is reference to original bug report and related work
- [ ] There is accuracy test, performance test and test data in opencv_extra repository, if applicable
      Patch to opencv_extra has the same branch name.
- [ ] The feature is well documented and sample code can be built with the project CMake

```
force_builders=Custom
buildworker:Custom=linux-4
Xbuild_image:Custom=ubuntu-cuda-cc52:18.04
build_image:Custom=ubuntu-cuda:16.04
```

### CODE REVIEW

File: modules/cudaimgproc/src/cuda/hough_lines.cu [hunk 1]
```diff
@@ -174,33 +171,30 @@ namespace cv { namespace cuda { namespace device
...
+            cudaSafeCall( cudaStreamSynchronize(stream) );
 
             totalCount = ::min(totalCount, maxSize);
 
             if (doSort && totalCount > 0)
             {
                 thrust::device_ptr<float2> outPtr(out);
                 thrust::device_ptr<int> votesPtr(votes);
-                thrust::sort_by_key(votesPtr, votesPtr + totalCount, outPtr, thrust::greater<int>());
+                thrust::sort_by_key(thrust::cuda::par.on(stream), votesPtr, votesPtr + totalCount, outPtr, thrust::greater<int>());
```
Comment (alalek) [inline modules/cudaimgproc/src/cuda/hough_lines.cu:197]: Could you please take a look if there is an easy way to guard this code for CUDA 8.x?

**Review by alalek — COMMENTED** [review]
Thank you for contribution!

File: modules/cudaimgproc/src/cuda/hough_lines.cu (hunk 1 above)
Comment (Atlas42) [reply modules/cudaimgproc/src/cuda/hough_lines.cu:197]: I do not understand your question here.
Is there a problem specific to CUDA 8 ? I have checked the API and it seems fine.
As Cuda 8 is not supported anymore on Ubuntu 18.04 or Visual Studio 19 (my dev tools),
I took some time today to setup an Ubuntu 16.04/CUDA 8 and it builds properly with OpenCV 4.5.0 (master did not on some unrelated problem).

Comment (alalek) [reply modules/cudaimgproc/src/cuda/hough_lines.cu:197]: Please take a look on build issues of "Custom" CUDA builder [here](http://pullrequest.opencv.org/#/summary/contrib).

There are several ways:
- try to fix the build using `#if` guards with CUDA version checks (requested above)
- try to fix the build through disabling this algorithm for CUDA 8.0
- open a discussion about deprecation of old CUDA versions in next OpenCV releases (will take some time)

Comment (Atlas42) [reply modules/cudaimgproc/src/cuda/hough_lines.cu:198]: It should be fixed on my last commit.
I managed to reproduce the build problem, the instruction on the NVIDIA website to install cuda 8 were wrong, I should have checked.
Turns out that the thrust::cuda::par did exist in CUDA 8, but was not included by default.

**Review by alalek — APPROVED** [review]
Well done! Thank you for contribution :+1:

**Pull request merged**